### DISCUSSION TITLE
`Dialog`: a more modular `Modal`

### DISCUSSION THREAD

**Discussion by chad1008** [WordPress/gutenberg#56231] [General]
@WordPress/gutenberg-components is looking into what the next version of `Modal` might look like, and we're considering (not a final decision by any means) potentially creating new component rather than simply rewriting the existing one. There are two main benefits to a new component:
1. We'd be able to give it a more meaningful/accurate name (`Dialog`)
2. We'd be able to follow a similar path to the ongoing `Tabs` work, where we are introducing composable sub-components, and also allowing the component to be used in both controlled and uncontrolled modes.

Our goal with this sub-component based approach is to create a component that is flexible rather than opinionated. It should semantically and accessibly implement the [WAI-ARIA Dialog pattern](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/). Or goal/recommendation is to create a `Dialog` that doesn't dictate what its contents should be, and then combine that with a set of relatively generic sub-components that can be used (or not used, as needed) by consumers.  

This approach would allow for more freedom and flexibility, so we'd like to gather feedback on what other improvements could be made over the current `Modal` component. What's missing from `Modal`? What's included in `Modal` but maybe doesn't work well? Or, are there any existing elements/behaviors of `Modal` that we don't really need?

Looking at some existing implementations of `Modal`, in the editor, the component often gets a lot of style overrides applied to it. This indicates the default design isn't up to date and/or isn't flexible enough. What are the most important features to improve that experience? 

One area in particular that design input would be appreciated: When reviewing the current API surface of `Modal` many, if not most, props could be eliminated using sub-components as children, making the component simpler overall. If we think of a new `Dialog` component as a collection of composable and/or optional sub-components, what would the ideal structure look like? What sub-components would be most useful?

One thing that is already on our radar is continuing to improve the preset sizing options in a way that gives consumers easy to use options, without becoming overly restrictive or opinionated. There have been a couple of different, but related conversations around this recently:
- https://github.com/WordPress/gutenberg/issues/55062
- https://github.com/WordPress/gutenberg/pull/55130

As work proceeds on this new component, we'll take those discussions into account and continue to iterate.

## Changes the new component would likely need

There are a couple of bigger changes that we do think would be involved in a new component, based on initial analysis.

First, `Modal` is rendered conditionally with the consumer managing the state that controls it. With `Dialog`, we'd use an `open` prop instead, so the implementation would be a little different than `Modal` in that regard.

Second, `Modal` has internal logic that's geared towards ensuring two non-nested modals cannot be opened at the same time. This isn't logic we think we'd like to carry into a new component, for a couple of reasons. First, rather than building a component that forces certain behaviors, we'd rather leave it flexible, and rely on consumers to render UIs properly and not trigger two modal dialogs at once. Also that's kind of tricky to do unless you really go out of your way to force it to happen.

There will likely be other differences that surface as work continues, but before going further into the project, we wanted to get input from others on what this new component will need most.

cc @WordPress/gutenberg-design  @richtabor @WordPress/gutenberg-components
(if you're reading this, please ping any other folks you think would have input to share!)

**Comment by jameskoster** [comment]
Should the Header and Footer be sub-components? It's always felt a little odd to me that there's a prop to _hide_ the header. This might enable a simplification (or even deprecation) of `ConfirmDialog`?

A dedicated footer component could allow for a more expressive design, so that we can do things like making the buttons fill the horizontal space on mobile:

<img width="375" alt="modal" src="https://github.com/WordPress/gutenberg/assets/846565/b5211f96-6f02-40b0-b3eb-f3688c977fac">

**Reply by ciampo** [reply]
Do we need dedicated subcomponents for header and footers? What would they have that is special vs using a simple `div`, or an `HStack` ? eg.

```jsx
<Dialog>
  <Dialog.Heading>Dialog title</Dialog.Heading>
  <p>Modal content</p>
  <HStack>
    <Dialog.DismissButton render={<Button variant="secondary" />}>Cancel</Dialog.DismissButton>
    <Dialog.DismissButton render={<Button variant="primary" />}>Confirm</Dialog.DismissButton>
  </HStack>
</Dialog>
```

My fear is that, the more "subcomponents" we have, the least flexible overall the component will be (which is what I think is already happening to `Modal`, judging by the many style overrides that it receives around the editor).

**Reply by jameskoster** [reply]
I do agree with the sentiment. For the header and footer I think the main argument would be about ensuring consistency. IE all titles use the correct type styles, footers contain at most 2 actions, the mobile behavior in the screenshot above... those kinds of things. Also if we change details down the road (let's say we want to add a separator between content and footer) all modals receive the updates at once.

**Reply by andrewhayward** [reply]
With a dedicated `<Dialog.Header>` and `<Dialog.Footer>` it would be a lot easier to ensure that content is consistently rendered.

For example, if we want to always have the title and actions visible...

<img src="https://github.com/WordPress/gutenberg/assets/159848/b35d0d38-a17f-4f7b-b353-5b086df75337" alt="A dialog window, with an 'Alert!' header and an action button reading 'Oh no...'. The content is scrollable." height="250px" align="left">

```jsx
<Dialog>
  <Dialog.Header title="Alert!">
    /* Any relevant controls... */
  </Dialog.Header>
  <p>Mauris blandit, arcu feugiat varius dapibus...</p>
  <Dialog.Footer>
    <Dialog.ActionButton
      render={<Button variant="primary" />}
      label="Oh no..."
      onClick={ ... }
    />
  </Dialog.Footer>
</Dialog>
```
<hr clear="left">

In general, I think I'd err on having additional sub-components if it makes consistency in the 90% of cases that are the same easier.

The `Header` might accept an `IconButton`, for example, but more often than not, it's going to be a standard `CloseButton`...

```jsx
<Dialog.HeaderIconButton icon={ <IconClose /> } align="end" label="Close" onClick={ ... } />
// or
<Dialog.HeaderCloseButton />
// which internally maps to the same thing, and automatically closes the dialog.
```

That doesn't stop consumers from doing this, if they still want to...

```jsx
<Dialog.HeaderCloseButton icon={ <IconBack /> } align="start" label="Back" />
<Dialog.HeaderIconButton icon={ <IconHelp /> } align="end" label="Help" onClick={ showHelp } />
```

Equally, in the footer, we might provide an `ActionButton`, but then also provide standardised versions for the typical use-cases...

```jsx
<Dialog.ActionButton render={<Button variant="secondary" />} label="Cancel" onClick={ ... } />
<Dialog.ActionButton render={<Button variant="primary" />} label="Confirm" onClick={ ... } />
// or
<Dialog.SecondaryActionButton label="Cancel" onClick={ ... } />
<Dialog.PrimaryActionButton label="Confirm" onClick={ ... } />
// or
<Dialog.ActionButton action={ Dialog.Actions.Cancel } />
<Dialog.ActionButton action={ Dialog.Actions.Confirm } />
// or
<Dialog.CancelActionButton />
<Dialog.ConfirmActionButton />
```

**Reply by ciampo** [reply]
@andrewhayward , In principle, I agree!

However, from personal experience, this has not always been the case:

> if we want to always have the title and actions visible...

I remember seeing usages of `Modal` where folks applied (hacky) overrides to change the size/spacing of the header, or introduce a sticky behaviour, etc etc. Basically, the requirements for specific use cases didn't fit what the component was allowing by default. And as soon as we have such overrides, it becomes very hard to make tweaks to the component itself, because they will likely break all of those scenarios where style overrides where applied.

Having said that, I think we could work towards a good compromise: having a set of header / footer sub-components that folks can decide to use in case they want to get maximum consistency. And in cases when the specific use-case requires a more custom layout, consumers can ditch those components and add their custom styles.

**Reply by ciampo** [reply]
@jameskoster 

> IE all titles use the correct type styles

For this specific functionality, a Heading subcomponent (ie. title) should be enough. A header subcomponent could be mostly useful for spacing and layout (eg. action/dismiss buttons rendered in the header subcomponent are automatically rendered in the top right corner)

> footers contain at most 2 actions

We would likely be able to implement this check strictly only if folks use the correct subcomponents (ie. Dialog.Footer and Dialog.FooterButton), but since we can't be guaranteed that folks won't use other custom components, we would probably at most be able to output a warning if we detect more than 2 `button`-like components as children.

Unless we want to have a super-strict footer element which doesn't allow custom `children` — but then we are again incurring the risk of having components that are not flexible enough

> if we change details down the road (let's say we want to add a separator between content and footer) all modals receive the updates at once.

This is where the risk of this approach lies, IMO. In an ideal world, where all consumers of this component use it as it comes out-of-the-box, making such changes would be extremely easy and effective.

But, out of experience, the reality in Gutenberg is that design needs change every week, and thus a component like this one would rapidly become unable to support all the different design specs. This means that folks would start applying style overrides, which ultimately make it really hard for us to make changes to the base component without breaking all those custom usecases.

So, as also mentioned in my reply to Andrew above, a middle ground would be to offer a set of subcomponents that are not excessively opinionated, and that optional for folks to use.

Not sure how everyone feels about it.

**Reply by chad1008** [reply]
The compromise approach of useful-but-not-overly opinionated sub-components has definite merits.

I agree with the value of easy ways to keep styling consistent. If we limit sub-component creation to just the things we expect most `Dialog`s to need (like a header/title), we can help consumers know they're doing things in a consistent manner. To Marco's point though, the moment one of those sub-components is too rigid or opinionated, it will get left out (which is fine) in favor of a custom alternative that might not maintain the same consistent styling over time.

So it kind of feels like we want sub-components we think would get used a lot. But we want to avoid having them do too much, so they stay applicable to as many use-cases as possible.

**Comment by jameskoster** [comment]
I suppose this is tied to the conversation around separating header / footer into sub-components, but this seems like a good moment to discuss `headerActions`. Having an action next to the close button has always felt a little unconventional to me, and after a quick search of the codebase it doesn't seem to be utilised anywhere currently. Should we deprecate this one?

**Reply by ciampo** [reply]
If we were to create a new component, `headerActions` (and any other render prop) wouldn't be required, because consumers of the component would just pass whatever children are needed to `Dialog`.

> Having an action next to the close button has always felt a little unconventional to me

I guess the main point of this conversation is to understand how much "design" we want to bake into `Dialog`. As also mentioned above, the more we want to enforce a certain design, the more consistency we will get across usages. At the same time, though, we the component will be less flexible to different needs and changes in design specs.

**Reply by jameskoster** [reply]
Yup it's a tricky one to balance. Consistency in design feels reasonably important to me, so I'd lean towards being more opinionated about the fundamental elements that 99% of modals use, IE sizing, header, footer, and general styling. I'm definitely open to being convinced otherwise though – not a strong feeling!

**Reply by chad1008** [reply]
> If we were to create a new component, headerActions (and any other render prop) wouldn't be required, because consumers of the component would just pass whatever children are needed to Dialog.

And looking at existing uses, I don't think it'll be missed. There may be third-party implementations that use it, but I don't actually see `headerActions` in use anywhere in gutenberg currently.

**Comment by chad1008** [comment]
## Initial Sub-Component Proposal

Looking at the above conversation, the key is going to be finding balance. Include simple sub-components that a high percentage of implementations will benefit from, but keep them simple so they're flexible and don't push consumers to start applying style overrides as they have to do with `Modal` today.

With that in mind, here's what I'm thinking, as a starting point. Feedback is very welcome, and we'll likely need to adjust and iterate on this list a bit as we go!

- `Dialog.Root`:
  - The main sub-component, responsible for rendering the `Dialog` itself. Accepts other sub-components/elements as children.
  - The only required sub-component.
  - Depending on how other sub-components are developed, might also serve as a context provider.
- `Dialog.Header`:
  - Must be wrapped in a `Dialog.Root` component to ensure `aria-labeledby` is properly applied to `Dialog.Root`.
  - Food for thought: if this component is not used, can we warn/require that consumers specify `aria-labeledby` explicitly?
  - Renders a title/heading with consistent styling.
  - Can also include a button to close the `Dialog`, toggled by a boolean prop. Undecided whether this button should be enabled or disabled by default, but it's a common enough pattern that I'd like to ensure consistency
  - The title and any other buttons/elements can be passed as children
- `Dialog.PrimaryButton`:
  - A consistent and easy to implement action button, primarily used for styling consistency
- `Dialog.DismissButton`:
  - A consistent and easy to implement dismiss button
  - Functionally the same as the optional dismiss button embedded in `Dialog.Header`


Notes:
- My hope is that providing two distinct button components (`Primary` and `Dismiss`) will not only ensure consistent styling, but also steer consumers towards not adding additional buttons. You shouldn't have two primary actions, and we're intentionally not offering you a "secondary" version. Hopefully that helps address the concern raised above around detecting and limiting/warning when too many buttons are provided?
- As work progresses, there may be some specific options/props/styles that are worth considering if they're common use-cases and don't make the sub-components too rigid.

**Reply by ciampo** [reply]
Some feedback:

- Dialog.Root will definitely need to hold some context and provide internally to the other subcomponents (for example, to coordinate dialog dismissal, dialog labelling etc)
- The component used to render a title for the Dialog (and accessibly label it) should be probably called `Dialog.Heading`.
- In addition, we could also have a `Dialog.Header` component which would act more as the wrapper, and could provide some styling/spacing/positioning/scroll behaviour? It would accept `children` (eg. `Dialog.Heading`), and/or render props (eg. `actions`?), and maybe by default show a dismiss button with the "X" icon?
- Similarly, we may want a `Dialog.Footer` component which ensures some spacing / styling etc
- Should the `DimissButton` always require children — ie. by default it doesn't render an X
- I get the reasons behind proposing a `PrimaryButton`, although I'm not sure that folks won't just render many of them anyway, or use other components to render differently styled buttons?
- As mentioned above, the tricky aspect is to balance how "opinionated" these subcomponents are:
  - If they're not very opinionated, one may argue that they don't offer much value and just shouldn't exist in the first place
  - If they're too opinionated, they risk being not flexible enough to accommodate the needs of the consumers, and they may just end up not being used

**Reply by andrewhayward** [reply]
Would be nice if `Dialog` could generate all of that for simple use-cases that don't need any custom tweaking.
That is, something like this...

```jsx
<Dialog title="Dialog Title" primaryActionLabel="Save" dismissActionLabel="Cancel">
  <p>This is the content of my dialog.</p>
</Dialog>
```

...would be functionally equivalent to...

```jsx
<Dialog.Root>
  <Dialog.Heading>Dialog Title</Dialog.Heading>
  <p>This is the content of my dialog.</p>
  <Dialog.Buttons> {/* 🤷‍♂️ */}
    <Dialog.DismissButton>Cancel</Dialog.DismissButton>
    <Dialog.PrimaryButton>Save</Dialog.PrimaryButton>
  </Dialog.Buttons>
</Dialog.Root>
```

...or whatever the API ends up looking like.

**Reply by ciampo** [reply]
> Would be nice if Dialog could generate all of that for simple use-cases that don't need any custom tweaking.

That sounds great until someone needs to use `Modal` in a slightly different way (maybe a different button style, different spacing around the header/footer, etc..), and they end up either passing a ton of style overrides, or not using the component at all.

Maybe a sweet spot could be to offer two versions of the component:
- A high-level, very opinionated version with a "monolithic" component approach (`<Dialog />`), similarly to the solution that Andrew just proposed in the message above
- A lower-level version that consumers can put together "a la carte" by picking and choosing subcomponents

**Reply by chad1008** [reply]
> Dialog.Root will definitely need to hold some context and provide internally to the other subcomponents (for example, to coordinate dialog dismissal, dialog labelling etc)

I've actually been impressed by how well Ariakit has handled some of those out of the box. Andrew's example above would actually work well based on my initial experiments. The internal context does a good job of handling things like labelling and dismissal. Unless we encounter situations where some of these subcomponents need to render in a `slot`/`fill` (like we found when implementing tabs) we may find we don't need our own context unless we have our own props that we want to share across sub-components.

> I get the reasons behind proposing a PrimaryButton, although I'm not sure that folks won't just render many of them anyway, or use other components to render differently styled buttons?

The button aspect of this is tricky! When I think about it, it feels like the risk of making things overly-opinionated is especially high with this part of the component.

> Would be nice if Dialog could generate all of that for simple use-cases that don't need any custom tweaking.
That is, something like this...

> Maybe a sweet spot could be to offer two versions of the component...

I think you're both on the same page here, unless I'm misreading, of having two versions: one monolithic and very simple to implement, the other composable and more flexible. I like that idea a lot, especially if we were to name them as Andrew described:

- `<Dialog {...} />` is a standalone component
- `<Dialog.abc>` is a namespaced group of sub-components

**Reply by andrewhayward** [reply]
> I think you're both on the same page here, unless I'm misreading, of having two versions: one monolithic and very simple to implement, the other composable and more flexible.

Yes, I'm pretty sure we're saying the same thing differently. The namespaced sub-components are (mostly?) unstyled primitives, which can be used explicitly by developers as needed. `<Dialog>` itself implicitly uses the same primitives, but in a more opinionated way.

We would essentially be exporting [Ariakit's dialog implementation](https://ariakit.org/components/dialog), with a nice wrapper around it for typical use-cases.

**Comment by ciampo** [comment]
This is a good example of what I refer to when I say that our current `Modal` component is too opinionated and not flexible enough:
https://github.com/WordPress/gutenberg/pull/53852#discussion_r1410442475

**Comment by afercia** [comment]
Chiming in here from https://github.com/WordPress/gutenberg/pull/53852#discussion_r1411057766

I'd agree making the modal dialog easier to use and based on sub-components would be welcome. That said, the `dialog` ARIA pattern _requires_ some features that can't be omitted. As such, allowing consumers to customize should be done _only to some extent_ because some features can't be removed.

Besides cosntrained tabbing, making th erest of the content inert and non perceivable, closing on focus and the like, there is one feature that is fundamental for accessibility: the dialog needs to be labeled.

We've had a few cases even in the editor where modal dialog were unlabelled. The new component must ensure the didalog is always labeled.

Ideally, for better accessibility, the dialog should be labelled by the means of _a visible H1 heading_ referenced with aria-labelledby on the element with `role=dialog`. We _could_ consider to optionally make the heading visually hidden but it must always be rendered. Since the heading typically lives in the dialog header, looks like the future `<Dialog.Header>` sub-component must be required. It should't be allowed to remove it.

The [dialog ARIA pattern](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/) requires that:

> The dialog has either:
> -A value set for the aria-labelledby property that refers to a visible dialog title.
> -A label specified by aria-label.

For the sake of simplification, I'd propose that the dialog is always labelled with the H1 and `aria-labelledby`. Optionally, we _could_ consider to provide the ability to visually hide the header but it should made clear that's not recommended.

Same applies to the default `Close` button. There must be a default way to close the modal that is not click outside and not the Escape key.

I think the most important thing to consider is that, while making the new component more flexible is reasonable, we also need to make sure WordPress provides accessible components _that are not open to misues_. Currently, some components are open to misues. For example, there have been several cases of unlabelled form controls and unlabelled modal dialogs. I'd really like all new components to avoid potential misues.

> Or goal/recommendation is to create a Dialog that doesn't dictate what its contents should be, and then combine that with a set of relatively generic sub-components that can be used (or not used, as needed) by consumers.

I'd agree but to some extent. The header / heading that provides the dialog labeling is arguably only 'content'. It is a required _functionak_ part of the dialog that shouldn't be optional.

> Looking at some existing implementations of Modal, in the editor, the component often gets a lot of style overrides applied to it. This indicates the default design isn't up to date and/or isn't flexible enough. 

I'd argue that design often doesn't understand the modal dialog does have _required_ features that can't be removed because, when removed, the accessibility level decreases. For example: styling the heading is perfectly legitimate. Visually hiding the heading is arguable and should be done only for very good reasons. Entirely removing the heading is a no-go.

**Reply by ciampo** [reply]
I understand the point of view and agree with some of the principles, but I'm not sure if there's a good way of achieving this at the code authoring level without specific APIs for accessing the Accessibility Tree Object Model.

For example, we could add a runtime check that either:
- the `Dialog.Heading` component is rendered inside the Modal, or
- the `Dialog` component itself has ah `h1` element and an `aria-labelledby` attribute pointing at it, or
- (maybe) the `Dialog` component itself has the `aria-label` attribute?

In case the check fails, we can output an error to the console. Although that would allow a consumer to still render the component.

Also, we'd need to deal with a lot of edge cases, eg. what if the `h1` has `display: none` set?


I'd be curious to know if and how `ariakit` deals with this @diegohaz

**Reply by diegohaz** [reply]
> I'd be curious to know if and how `ariakit` deals with this @diegohaz

Ariakit has a `DialogHeading` component that automatically assigns the `aria-labelledby` attribute to the `Dialog`. However, the application or design system is responsible for enforcing its use. Implementing this check in a broad manner is far from simple and significantly affects runtime performance and bundle size, even if the check is executed only during development. The app or design system can carry out this check more efficiently with more specific domain knowledge.

> what if the `h1` has `display: none` set?

A `<h1 hidden>` or `<h1 style="display:none">` element remains visible to the accessibility tree if it's referenced by the `aria-labelledby` attribute.

If the `DialogDismiss` component isn't rendered, Ariakit automatically renders a visually hidden dismiss button if the dialog is modal.

**Reply by diegohaz** [reply]
One option is to enforce the use of a `title` attribute or similar, but allow a `dangerously*` prop to overwrite/customize it.

**Reply by ciampo** [reply]
> However, the application or design system is responsible for enforcing its use. Implementing this check in a broad manner is far from simple and significantly affects runtime performance and bundle size, even if the check is executed only during development. The app or design system can carry out this check more efficiently with more specific domain knowledge.

Thank you for sharing your thoughts. This has also been my stance around similarly related issues for the very same reasons, and I'm not sure exactly how a components library can successfully be _"not open to misuse"_ as @afercia suggested above.

In this specific case, an alternative would be, as Diego suggested, to have a required `title` prop; although:

- folks can still easily decide not to pass one and disable linter errors / ignore console errors
- it would completely defeat the point of allowing for more modularity

**Reply by andrewhayward** [reply]
Couple of thoughts...

> Besides constrained tabbing, making the rest of the content inert and non perceivable, closing on focus and the like...

As an aside, these are valid points, but it's worth noting that this is intentionally not a `<ModalDialog>`. So while we should make sure these aspects are considered, they are not core requirements, but rather implementation details around any `modal` prop (or similar).

> ...accessible components that are not open to misuse...

While I entirely understand the idea, from a technical perspective, this is not possible. There will always be ways of misusing components – we can require labels and descriptions, for example, and (though expensive) do runtime checks to assert constraints, but we can never validate that they are correct, relevant or useful.

Of course, that is not to say that we shouldn't do our best to construct APIs that make correct usage easier, and minimise any potential issues. It might be reasonable to require `Dialog.Header` if, for example, a `title` prop isn't set on the `Dialog.Container`. My hypothetical TypeScript isn't great, but something like this...

```typescript
interface Header {
    title: string
    // ...
};

interface Footer {
    // ...
};

type Child = { id: string };

type Children = [...Child[], Footer | Child];

interface DialogBase {
    // ...
}

interface DialogWithTitle extends DialogBase {
    title: string,
    children?: Children,
};

interface DialogWithHeader extends DialogBase {
    title?: never,
    children: [Header, ...Children],
}

type Dialog = DialogWithTitle | DialogWithHeader;
```

**Comment by ciampo** [comment]
> First, `Modal` is rendered conditionally with the consumer managing the state that controls it. With `Dialog`, we'd use an `open` prop instead, so the implementation would be a little different than `Modal` in that regard.

As also discussed in https://github.com/WordPress/gutenberg/pull/65203, switching from conditionally rendering the component to having an `open` prop would also enable exit animations when closing the dialog.

**Comment by ciampo** [comment]
Closing this conversation as [a new `Dialog` component has been added to the `@wordpress/ui` pacakge](https://github.com/WordPress/gutenberg/pull/75183). Thank you all for contributing to shaping its APIs.
